Make shared library produced by cmake and configure identical

dankegel opened the pull request:
zlib shared libraries really ought to be identical regardless of whether you build them with cmake or configure.
This pull request verifies that, and makes it so, on four platforms, at least for plain old and --zlib-compat builds.

Here's what it took (beyond the static library fix landed recently):
- canonicalize CHOST in configure
- rearrange source file order in cmake and configure to match (at least for i686, amd64, aarch64, ppc, and ppc64le)
- don't link shared libraries explicitly with -lc (shouldn't be needed; caused differences between configure and cmake)
- on mac, do like configure, and don't use @rpath
- update pkgcheck.sh to verify that cmake and configure create identical shared libraries

Along the way, it turned out that the configure sections for 32 and 64 bit x86 were
essentially identical, so this pull request merges them to reduce chance of
accidental divergence later.

Fixes #638 and #666 

Known issues left for future pull requests:
- There may still be differences between cmake and configure on architectures not yet tested by github/workflows/pkgcheck.yaml
- --native and -DWITH_NATIVE_INSTRUCTIONS don't do quite the same thing (see #697)

dankegel committed:
configure: don't link shared libraries explicitly with -lc

The -lc was introduced by d6231142d2b883a8c3b253fa34992b5cdb4ac2fe :
which said in part
"Changes in 1.2.3.2 (3 September 2006)
...
- Rig configure --shared to build both shared and static [Teredesai, Truta]"

It's not clear why it was needed.  On powerpc, linking -lc explicitly
causes __moddi3 and __stack_chk_fail_local to swap location in the
binary.
If we want configure and cmake to generate identical shared libraries,
either we need to link -lc explicitly in both configure and cmake,
or in neither.   Occam's Razor suggests omitting it from configure
rather than adding it to cmake.

dankegel commented:
Rebased.

dankegel commented:
This is ready to go, I think.

nmoinvaz commented:
I think the problem with configure script for PPC detection when cross compile should be fixed.

And configure and cmake should be set to -O2. 

That should fix some workarounds that were needed I believe, also you won’t need to check CMakeCache.txt.

Dead2 commented on configure:
```diff
@@ -1173,11 +1173,22 @@ case "${ARCH}" in
+
+            # if(WITH_SSE4 AND (HAVE_SSE42CRC_INLINE_ASM OR HAVE_SSE42CRC_INTRIN))
+            CFLAGS="${CFLAGS} -DX86_CPUID -DX86_SSE42_CRC_HASH"
+            SFLAGS="${SFLAGS} -DX86_CPUID -DX86_SSE42_CRC_HASH"
```
Duplicates -DX86_CPUID from the above block?

dankegel replied:
Thanks.

Dead2 commented on test/pkgcheck.sh:
```diff
@@ -40,24 +100,59 @@ cd btmp1
+  export CFLAGS
+  # Hack: oddly, given CHOST=powerpc-linux-gnu, configure concludes uname is gnu,
+  # causing it to set LDSHAREDFLAGS to a different value than cmake uses.
+  # So override it when appropriate.  FIXME
```
Perhaps fix this powerpc-detection in configure instead?

mscastanho replied:
I believe the issue is on `configure:26`:
https://github.com/zlib-ng/zlib-ng/blob/9abe04b926ef688efa3d4f464209246a22afb01d/configure#L25-L28
When `CHOST` is something like `powerpc-linux-gnu`, it will match the second pattern, which will select the last string after the `-`. But this issue happens with any `CHOST` value in the form `first-second-third`, such as `arm-linux-gnueabihf` and `aarch64-linux-gnu`. All these match the second pattern.

That change may deserve a separate PR though. For now I believe setting `--uname=linux` will suffice.

mscastanho replied:
Just opened a new issue about this so it doesn't get forgotten: https://github.com/zlib-ng/zlib-ng/issues/666

dankegel replied:
I added a commit to address this.

mscastanho commented on test/pkgcheck.sh:
```diff
@@ -40,21 +43,35 @@ cd btmp1
+  then
+    echo "Warning: libz$suffix.a does not match.  Assuming ar needs -D option.  Unpacking..."
+    cd pkgtmp1; ar x usr/local/lib/libz$suffix.a; rm usr/local/lib/libz$suffix.a; cd ..
+    cd pkgtmp2; ar x usr/local/lib/libz$suffix.a; rm usr/local/lib/libz$suffix.a; for a in *.c.o; do b=$(echo $a | sed 's/\..*//'); mv $a $b.o; done; cd ..
```
The inner part of that for loop could be simplified to just `mv $a ${a%%.*}.o`. But to be honest I'm not sure if that's portable to other shells too. If not, just disconsider.

dankegel replied:
That syntax looks strange to me :-)  But yes, that seems portable and would be shorter.  If I get a vote or two 
for the %% syntax, I'll switch to it.

mscastanho commented on .github/workflows/pkgcheck.yml:
```diff
@@ -8,14 +8,30 @@ jobs:
+            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-aarch64.cmake
+            packages: ninja-build qemu gcc-aarch64-linux-gnu libc-dev-arm64-cross
+
+          - name: Ubuntu GCC PPC
```
I believe we should switch this to PPC64LE. This is the ABI followed by recent POWER processors and I believe not many recent distros still support ppc or ppc64, so if we need to choose one, ppc64le is the way to go since it will cover most of the current ppc user base.

dankegel replied:
Happy to add ppc64le, but isn't a lot of embedded powerpc going to be 32 bit for the foreseeable future?  So maybe keep the 32 bit one... I have a Powermac G4, and I'm biased :-)

mscastanho replied:
Agreed. The more the merrier =)

mscastanho commented on test/pkgcheck.sh:
```diff
@@ -1,10 +1,67 @@
+$ export CHOST=aarch64-linux-gnu
+$ export CMAKE_ARGS="-DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-aarch64.cmake -DCMAKE_C_COMPILER_TARGET=${CHOST}"
+
+ppc:
```
Some as I commented above. Probably better to use ppc64le.

dankegel replied:
Added.

mscastanho commented on configure:
```diff
@@ -69,7 +69,7 @@ fi
 LDCONFIG=${LDCONFIG-"ldconfig"}
 LDFLAGS=${LDFLAGS-"-L."}
-LDSHAREDLIBC="${LDSHAREDLIBC--lc}"
+LDSHAREDLIBC="${LDSHAREDLIBC}"
```
Looks good. If somebody actually needs `-lc` they can just set it before calling configure.

dankegel replied:
Agreed.

mscastanho reviewed (commented):
Some comments here and there but overall the PR looks good. My main suggestion is to use ppc64le instead of ppc in the CI tests.

mscastanho commented:
Also an additional comment in the `Usage` string of `test/pkgcheck.sh` saying that the example commands are for Ubuntu/Debian-like distros would be nice. Since the instructions use `apt` and the names of the dependency packages may change on other distros. Something simple like `For example, on Ubuntu:`.

dankegel committed:
configure: canonicalize CHOST.

Fixes https://github.com/zlib-ng/zlib-ng/issues/666
for all the CHOST values currently used in
.github/workflows/*.yml.

Also make ci's shellcheck happy by avoiding backticks in the changed lines.

dankegel commented:
Resolved except maybe for the %% and the "on ubuntu" wording in test/pkgcheck.sh comments.
Tomorrow morning I'll push an update for those and anything else that comes up tonight.

nmoinvaz commented:
@dankegel have you encountered the issue with `-mfpu=neon` being set on all files when Neon is present in configure, but in CMake we only set it per file that uses neon?

dankegel commented:
Yes.  That's why I wimped out and only test aarch64, not 32 bit arm...

nmoinvaz commented:
See #669 and #668. Maybe you can cherry pick those and see if they solve the problems?

mscastanho commented on .github/workflows/pkgcheck.yml:
```diff
@@ -8,14 +8,37 @@ jobs:
+            chost: powerpc64le-linux-gnu
+            compiler: powerpc64le-linux-gnu-gcc
+            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-powerpc64le.cmake
+            packages: ninja-build qemu gcc-powerpc64le-linux-gnu libc6-dev-ppc64-powerpc-cross
```
@dankegel I believe the glibc package you need here is actually `libc6-dev-ppc64el-cross`

dankegel replied:
thanks.  yesterday was not my best day :-)

mscastanho commented on .github/workflows/pkgcheck.yml:
```diff
@@ -8,14 +8,37 @@ jobs:
+            chost: powerpc-linux-gnu
+            compiler: powerpc-linux-gnu-gcc
+            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-powerpc.cmake
+            packages: ninja-build qemu gcc-powerpc-linux-gnu libc-dev-powerpc-cross
```
glibc packages on Ubuntu seem to start with `libc6`. So this should be `libc6-dev-powerpc-cross`

dankegel replied:
Yes, looks like the libc- one is an alias, but it's always less confusing to use the real package name.

mscastanho commented on .github/workflows/pkgcheck.yml:
```diff
@@ -8,14 +8,37 @@ jobs:
+            chost: aarch64-linux-gnu
+            compiler: aarch64-linux-gnu-gcc
+            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-aarch64.cmake
+            packages: ninja-build qemu gcc-aarch64-linux-gnu libc-dev-arm64-cross
```
`libc6-dev-arm64-cross` here?

dankegel replied:
I used same package name as in cmake.yml.  Evidently the one missing the 6 is an alias, and it pulls in the right package.  Both files probably should use the real package name.

mscastanho commented on .github/workflows/pkgcheck.yml:
```diff
@@ -8,14 +8,37 @@ jobs:
-          Ubuntu GCC
+          Ubuntu GCC,
+          Ubuntu GCC AARCH64,
+          Ubuntu GCC PPC
```
@dankegel Shouldn't you add `Ubuntu GCC PPC64LE` to this list too?

dankegel replied:
yep.

mscastanho commented on test/pkgcheck.sh:
```diff
@@ -1,10 +1,72 @@
+$ export CMAKE_ARGS="-DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-powerpc.cmake"
+
+ppc64le:
+$ sudo apt install ninja-build diffoscope qemu gcc-powerpc64le-linux-gnu libc6-dev-ppc64-powerpc-cross
```
Should be `libc6-dev-ppc64el-cross`. And also replace `libc` by `libc6` on the other package names.

mscastanho reviewed (commented):
@dankegel just some small comments about package names. I think the builds passed because the glibc cross packages are in the `recommends` list for the gcc package, but having the correct package names is ideal.
Also, your fix for the `CHOST` issue looks good. Now `uname` is correctly being detected as `linux` for those failing cases.

dankegel commented:
cherry-picking #668 and #669 still leaves us with configure applying -march=armv8-a+crc hither and yon for
arm-linux-gnueabihf, e.g. on adler32.c.

nmoinvaz commented:
@dankegel are there also problems with configure using powerpc? I noticed we don't have a CI test for powerpc for configure..

dankegel commented:
powerpc seems ok.  arm is more attached to -march flags than powerpc is, I think...

dankegel commented:
The cmake O2/O3 issue seems beyond the goals of this pull request; my strategy of letting the wookie, er cmake, win seems like enough for the moment.

And I'm not sure this pull request should be blocked waiting for the 32 bit -march disagreement between cmake and configure to be solved...?  A later pull request can expand the list of architectures tested by pkgcheck.

Other than those, I think the pull request now covers the feedback received, and it's ready for merging.

mtl1979 commented on configure:
```diff
@@ -1173,11 +1175,22 @@ case "${ARCH}" in
 
-            ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} x86.o chunkset_sse.o insert_string_sse.o slide_sse.o"
-            ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} x86.lo chunkset_sse.lo insert_string_sse.lo slide_sse.lo"
+            # if(WITH_SSE4 AND (HAVE_SSE42CRC_INLINE_ASM OR HAVE_SSE42CRC_INTRIN))
```
Short-circuiting the test should be reverted...

mtl1979 commented on configure:
```diff
@@ -1173,11 +1175,22 @@ case "${ARCH}" in
-            ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} x86.o chunkset_sse.o insert_string_sse.o slide_sse.o"
-            ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} x86.lo chunkset_sse.lo insert_string_sse.lo slide_sse.lo"
+            # if(WITH_SSE4 AND (HAVE_SSE42CRC_INLINE_ASM OR HAVE_SSE42CRC_INTRIN))
+            CFLAGS="${CFLAGS} -DX86_SSE42_CRC_HASH"
```
Assuming SSE4.2 is available is still bad as a lot of AMD processors don't have SSE4.2 as it was originally available on only Intel processors. I don't know when the test was removed as even my previous development machine didn't have SSE4.2 and zlib-ng worked on it using configure...

dankegel replied:
I think there was never a check in configure; I pulled in the check from CMakeLists.txt and
added it as a comment in configure.  I admit that's a bit confusing.  I suspect zlib-ng checks
for sse4.2 at runtime before using it...?

nmoinvaz commented:
I think we should wait until some of the cmake/configure issues are resolved and then add more configurations to the pkgcheck. Sometimes these things take time. I have had PRs that take 3-4 months so I know how it is.

dankegel commented:
h'ok.  Landing the other three pull requests first is fine.

dankegel committed:
CMakeLists.txt, configure: match source order; .so's now identical on more platforms.

Also:
- use same name for .so on CMake and configure in --zlib-compat case
- configure: merge i686 and x86_64 sections, since they were essentially identical
- pkgcheck.sh: verify resulting shared libraries are bit-for-bit identical
- pkgcheck.sh: use diffoscope if present to show differences in shared libraries

Don't install diffoscope in ci, since that slows down successful
runs; we can add it later if failures become common.

Fixes #638 (at least for the systems tested)

dankegel commented:
Let's do this :-)

mtl1979 commented:
This overlaps with #676 ?

dankegel commented:
Only trivially.  git will do the right thing.  (I think.)

In retrospect, I suppose I should have waited until #676 was merged before saying "let's do this", to avoid that question.

(Merge could happen in either order, but if you merge this one first, #676 would be a no-op.)

mtl1979 commented:
As long as @Dead2 knows the correct order of merging pull requests...

nmoinvaz commented:
@dankegel thank you for adding the new architectures. Are there any remaining outstanding issues that we haven't cataloged? I had also noticed the issue with `LDSHAREDLIBC` too during test.  And I also see that the line `LDFLAGS=${LDFLAGS-"-L."}` adds `-L.` which is not added in CMake.

nmoinvaz commented on config.sub:
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
Is it possible to do this inline in `configure` without a separate file?

dankegel replied:
Yes, but config.sub is traditional, and if somebody needs one of the arches not yet in this config.sub,
they can drop in the gnu one and get unblocked.  So there's utility in keeping it a separate file.

nmoinvaz replied:
Is it possible to merge this file into the configure script?

nmoinvaz replied:
Is it possible to merge this file into the configure script?

nmoinvaz commented on config.sub:
```diff
@@ -0,0 +1,13 @@
+# https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub
+# but accepts a much smaller set of machine specifications,
+# and is under the zlib license rather than the GPL.
+# See also https://bugs.gentoo.org/651908
```
I don't know that we need to mention GPL here.

dankegel replied:
Yeah, that could confuse license scanners.

nmoinvaz commented on test/pkgcheck.sh:
```diff
@@ -1,9 +1,81 @@
+
+# Tell GNU's ld etc. to use Jan 1 1970 when embedding timestamps
+# Probably only needed on older systems (ubuntu 14.04?)
+export SOURCE_DATE_EPOCH=0
```
Do we need this `SOURCE_DATE_EPOCH`? Maybe not..?

dankegel replied:
I believe I needed it on FreeBSD 12.1, Debian 8, and Ubuntu 14.04. 

See also https://reproducible-builds.org/docs/source-date-epoch/ for what this does.

nmoinvaz commented on test/pkgcheck.sh:
```diff
@@ -40,24 +112,55 @@ cd btmp1
+Darwin)
+  # Alas, dylibs still have an embedded hash or something,
+  # so nuke it.
+  # FIXME: find a less fragile way to deal with this.
```
Is there something we can do about this? How many bytes were changed?

dankegel replied:
Per the dd command, 16 bytes.

One does not do anything about Apple tools, one endures them :-)

mtl1979 replied:
Would be nice to have some link to site that describes what those bytes actually are... Mach-O file format reference doesn't have clear definition what starts at byte offset 1337 as the file header isn't that long...

dankegel replied:
Yes, that would be nice.  However, I don't know how to dig it up myself, and I don't know that it's worth holding up this pull request for.

nmoinvaz replied:
I'm not sure if this is current but `strip` seems to have a `-no_uuid` flag that will strip out all `LC_UUID` load commands.
https://www.manpagez.com/man/1/strip/

dankegel replied:
That sounds handy.  Also, evidently ld has a -no_uuid option.  I'll see if those can replace the dd!

dankegel replied:
So far no joy.  It was (and maybe is) a good theory, though.

I just refreshed the mac support a bit, turned up an interesting bit so I left it as a separate commit.

dankegel replied:
Wait, there is joy in mudville.  Stand by...

dankegel replied:
I got it to go green, but there's a catch.  See https://github.com/zlib-ng/zlib-ng/issues/681

nmoinvaz reviewed (commented):
I have given some feedback on a few things.

dankegel commented:
-L. appears to do nothing and could be omitted.  I looked through git history and found it may have originally been
added in f6194ef3, but even then, it was a no-op (unless there is some strange ld out there
that searches the library search path even when you provide the filename of the library).

Likewise, the -L.. could be omitted from test/Makefile.in.

But as it does not cause configure and cmake to generate different bits,
I hesitate to remove it as part of this pull request... perhaps it could be a followon.

dankegel commented:
OK, I updated config.sub's comment to explain what it does better, and to
mention using autotools' config.sub as a workaround in case this one is too limited.

I also rebased.

I think that (and my replies above) resolves all the feedback received today sufficiently.
If the magic dd for macos seems unacceptable, I can remove macos support from pkgcheck.sh.

Please merge #676 first, then I'll rebase.

nmoinvaz commented on test/pkgcheck.sh:
```diff
@@ -40,24 +118,57 @@ cd btmp1
-    ;;
   esac
-  ../configure
+  # Use same optimization level as cmake did.
```
We probably won’t need this after #674 is merged.

dankegel replied:
Gone.

dankegel commented:
h'ok, it passes on mac now, too, without the dastardly dd hack.

How do people feel about the way the mac AVX2 problem was
solved by overloading WITH_NATIVE_INSTRUCTIONS ?
I could have introduced a new option, but WITH_NATIVE_INSTRUCTIONS
is such a close match.

dankegel commented:
Is that code coverage decrease real?

Dead2 commented:
This is now ready for rebase.

mscastanho commented:
@dankegel I noticed that this leaves directories `btmp1`, `btmp2`, `pkgtmp1` and `pkgtmp2` after running the script. Maybe add them to `make clean`? Another option would be to erase them after the check is done, but it might be useful to be able to check them in case of error. This could easily be part of a future clean up PR as well.

dankegel commented:
@mscastanho I added clean and .gitignore entries for pkgcheck's detritus, and made pkgcheck clean up after itself except on failure.

dankegel commented:
Rebased and ready for final review and landing.

dankegel commented:
The coverage changes seem spurious, but they're caused by something, and this pull request probably has to wait until it's understood.  It's mildly annoying that they only show up in the CI of the pull request, not the CI of the branch before creating the pull request...

dankegel commented:
Let's land #692 first, then I'll rebase and figure out that coverage problem.

dankegel committed:
cmake: mac: don't use @rpath (since configure doesn't); for #638

nmoinvaz commented on configure:
```diff
@@ -1118,8 +1120,8 @@ ARCH_SHARED_OBJS=''
-    # x86 specific optimizations
-    i386 | i486 | i586 | i686)
+    # x86/amd64 specific optimizations
+    i386 | i486 | i586 | i686 |x86_64)
```
I assume with these changes now we always do SSE2 check for x64 instead of assuming it exists (like CMake) and that you have merged the x86 and x64 sections in configure?

dankegel replied:
Yup.

nmoinvaz reviewed (commented):
I have made only two comments. It looks much smaller indeed!

nmoinvaz commented:
It seems I already asked the question about config.sub. I think it would be better if it was merged into configure in some way. That way configure script is more self contained. 

An alternative might be to create a “configure” directory same as the “CMake” directory and put it there but I’m not that fond of that idea either.

dankegel committed:
configure: move config.sub to tools subdir

dankegel commented:
I removed the questionable set of CFLAGS in test/pkgcheck.sh, and moved config.sub out of the top level directory.

I think it really is good to go now...?

nmoinvaz reviewed (approved):
I think this commit makes the code better - I like how it no longer checks the CMakeCache.txt. Having the file in the _tools_ directory is a nicer too. @Dead2 will be able to speak to if he wants it there or in a _configure_ directory. Otherwise, LGTM

dankegel commented:
Can't have it in a 'configure' subdirectory, there's a bit of a name clash there :-)

dankegel commented:
Push the button @Dead2 :-)

Dead2 commented:
I am in no condition to review stuff right now, with 5 holes in my abdomen and recovering from surgery that did not go quite as planned (it turned out OK, but means a much longer and more painful recovery). I hope to be able to read without my eyes constantly crossing in a couple of days, after that I can probably get some more testing and reviewing done.

That said, this PR has been reviewed multiple times by me and others, and although I see only 1 current review approval, I'll merge this now so we can keep things moving along and get the rest rebased on top of this as well. The other PRs will have to wait a bit longer though, unless there are more-or-less trivial changes that get at least TWO review approvals in addition to me.

dankegel commented:
Thank you and get well soon!
